File: rwanda_locations.py
"""
Rwanda Locations - Easy Integration Library

A simple, powerful utility to work with Rwanda's administrative locations.
Supports: Provinces, Districts, Sectors, Cells, and Villages

For a more feature-rich package version, visit:
https://github.com/DevRW/rwanda-location
"""

import json
import os
from typing import List, Dict, Optional, Any


class RwandaLocations:
    def __init__(self, data_file: str = None):
        """
        Initialize the Rwanda Locations library

        Args:
            data_file: Path to the locations.json file (default: locations.json in the same directory)
        """
        self.data_file = data_file or os.path.join(os.path.dirname(__file__), 'locations.json')
        self.data: List[Dict] = None
        self.cache = {
            'provinces': None,
            'districts': None,
            'sectors': None,
            'cells': None,
            'villages': None
        }

    def load(self) -> 'RwandaLocations':
        """Load the locations data from JSON file"""
        if self.data is None:
            with open(self.data_file, 'r', encoding='utf-8') as f:
                self.data = json.load(f)
        return self

    def get_provinces(self) -> List[Dict[str, Any]]:
        """
        Get all provinces

        Returns:
            List of unique provinces with code and name
        """
        if self.cache['provinces']:
            return self.cache['provinces']

        self.load()
        provinces = {}

        for location in self.data:
            code = location['province_code']
            if code not in provinces:
                provinces[code] = {
                    'code': location['province_code'],
                    'name': location['province_name']
                }

        self.cache['provinces'] = sorted(provinces.values(), key=lambda x: x['code'])
        return self.cache['provinces']

    def get_districts(self, province_code: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get all districts, optionally filtered by province

        Args:
            province_code: Optional province code to filter by

        Returns:
            List of districts
        """
        self.load()
        districts = {}

        for location in self.data:
            if province_code is not None and location['province_code'] != province_code:
                continue

            code = location['district_code']
            if code not in districts:
                districts[code] = {
                    'code': location['district_code'],
                    'name': location['district_name'],
                    'province_code': location['province_code'],
                    'province_name': location['province_name']
                }

        return sorted(districts.values(), key=lambda x: x['code'])

    def get_sectors(self, district_code: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get all sectors, optionally filtered by district

        Args:
            district_code: Optional district code to filter by

        Returns:
            List of sectors
        """
        self.load()
        sectors = {}

        for location in self.data:
            if district_code is not None and location['district_code'] != district_code:
                continue

            code = location['sector_code']
            if code not in sectors:
                sectors[code] = {
                    'code': location['sector_code'],
                    'name': location['sector_name'],
                    'district_code': location['district_code'],
                    'district_name': location['district_name'],
                    'province_code': location['province_code'],
                    'province_name': location['province_name']
                }

        return sorted(sectors.values(), key=lambda x: x['code'])

    def get_cells(self, sector_code: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get all cells, optionally filtered by sector

        Args:
            sector_code: Optional sector code to filter by

        Returns:
            List of cells
        """
        self.load()
        cells = {}

        for location in self.data:
            if sector_code is not None and location['sector_code'] != sector_code:
                continue

            code = location['cell_code']
            if code not in cells:
                cells[code] = {
                    'code': location['cell_code'],
                    'name': location['cell_name'],
                    'sector_code': location['sector_code'],
                    'sector_name': location['sector_name'],
                    'district_code': location['district_code'],
                    'district_name': location['district_name'],
                    'province_code': location['province_code'],
                    'province_name': location['province_name']
                }

        return sorted(cells.values(), key=lambda x: x['code'])

    def get_villages(self, cell_code: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get all villages, optionally filtered by cell

        Args:
            cell_code: Optional cell code to filter by

        Returns:
            List of villages
        """
        self.load()
        villages = {}

        for location in self.data:
            if cell_code is not None and location['cell_code'] != cell_code:
                continue

            code = location['village_code']
            if code not in villages:
                villages[code] = {
                    'code': location['village_code'],
                    'name': location['village_name'],
                    'cell_code': location['cell_code'],
                    'cell_name': location['cell_name'],
                    'sector_code': location['sector_code'],
                    'sector_name': location['sector_name'],
                    'district_code': location['district_code'],
                    'district_name': location['district_name'],
                    'province_code': location['province_code'],
                    'province_name': location['province_name']
                }

        return sorted(villages.values(), key=lambda x: x['code'])

    def search(self, search_term: str, level: str = 'all') -> List[Dict[str, Any]]:
        """
        Search locations by name (case-insensitive)

        Args:
            search_term: Search term
            level: Level to search: 'province', 'district', 'sector', 'cell', 'village', or 'all'

        Returns:
            List of matching locations
        """
        self.load()
        term = search_term.lower()
        results = []
        seen = set()

        for location in self.data:
            matches = (
                (level == 'all' or level == 'province') and term in location['province_name'].lower() or
                (level == 'all' or level == 'district') and term in location['district_name'].lower() or
                (level == 'all' or level == 'sector') and term in location['sector_name'].lower() or
                (level == 'all' or level == 'cell') and term in location['cell_name'].lower() or
                (level == 'all' or level == 'village') and term in location['village_name'].lower()
            )

            if matches:
                key = f"{location['province_code']}-{location['district_code']}-{location['sector_code']}-{location['cell_code']}-{location['village_code']}"
                if key not in seen:
                    seen.add(key)
                    results.append({
                        'province': {'code': location['province_code'], 'name': location['province_name']},
                        'district': {'code': location['district_code'], 'name': location['district_name']},
                        'sector': {'code': location['sector_code'], 'name': location['sector_name']},
                        'cell': {'code': location['cell_code'], 'name': location['cell_name']},
                        'village': {'code': location['village_code'], 'name': location['village_name']}
                    })

        return results

    def get_location_by_village_code(self, village_code: int) -> Optional[Dict[str, Any]]:
        """
        Get location hierarchy for a specific village

        Args:
            village_code: Village code

        Returns:
            Complete location hierarchy or None if not found
        """
        self.load()

        for location in self.data:
            if location['village_code'] == village_code:
                return {
                    'country': {'code': location['country_code'], 'name': location['country_name']},
                    'province': {'code': location['province_code'], 'name': location['province_name']},
                    'district': {'code': location['district_code'], 'name': location['district_name']},
                    'sector': {'code': location['sector_code'], 'name': location['sector_name']},
                    'cell': {'code': location['cell_code'], 'name': location['cell_name']},
                    'village': {'code': location['village_code'], 'name': location['village_name']}
                }

        return None

    def get_hierarchy(self, province_code: Optional[int] = None,
                     district_code: Optional[int] = None,
                     sector_code: Optional[str] = None,
                     cell_code: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get hierarchy tree structure

        Args:
            province_code: Optional province code to filter by
            district_code: Optional district code to filter by
            sector_code: Optional sector code to filter by
            cell_code: Optional cell code to filter by

        Returns:
            Hierarchical tree structure
        """
        self.load()
        tree = {}

        for location in self.data:
            if province_code is not None and location['province_code'] != province_code:
                continue
            if district_code is not None and location['district_code'] != district_code:
                continue
            if sector_code is not None and location['sector_code'] != sector_code:
                continue
            if cell_code is not None and location['cell_code'] != cell_code:
                continue

            p_key = location['province_code']
            d_key = location['district_code']
            s_key = location['sector_code']
            c_key = location['cell_code']

            if p_key not in tree:
                tree[p_key] = {
                    'code': location['province_code'],
                    'name': location['province_name'],
                    'districts': {}
                }

            if d_key not in tree[p_key]['districts']:
                tree[p_key]['districts'][d_key] = {
                    'code': location['district_code'],
                    'name': location['district_name'],
                    'sectors': {}
                }

            if s_key not in tree[p_key]['districts'][d_key]['sectors']:
                tree[p_key]['districts'][d_key]['sectors'][s_key] = {
                    'code': location['sector_code'],
                    'name': location['sector_name'],
                    'cells': {}
                }

            if c_key not in tree[p_key]['districts'][d_key]['sectors'][s_key]['cells']:
                tree[p_key]['districts'][d_key]['sectors'][s_key]['cells'][c_key] = {
                    'code': location['cell_code'],
                    'name': location['cell_name'],
                    'villages': []
                }

            villages = tree[p_key]['districts'][d_key]['sectors'][s_key]['cells'][c_key]['villages']
            if not any(v['code'] == location['village_code'] for v in villages):
                villages.append({
                    'code': location['village_code'],
                    'name': location['village_name']
                })

        # Convert nested dicts to lists
        result = []
        for province in tree.values():
            province['districts'] = [
                {
                    **district,
                    'sectors': [
                        {
                            **sector,
                            'cells': list(sector['cells'].values())
                        }
                        for sector in district['sectors'].values()
                    ]
                }
                for district in province['districts'].values()
            ]
            result.append(province)

        return result

    def get_stats(self) -> Dict[str, int]:
        """
        Get statistics about the data

        Returns:
            Dictionary with counts of each administrative level
        """
        self.load()

        return {
            'total_locations': len(self.data),
            'provinces': len(self.get_provinces()),
            'districts': len(self.get_districts()),
            'sectors': len(self.get_sectors()),
            'cells': len(self.get_cells()),
            'villages': len(self.get_villages())
        }


# Example usage
if __name__ == '__main__':
    # Create instance
    locations = RwandaLocations()

    # Get all provinces
    provinces = locations.get_provinces()
    print(f"Total Provinces: {len(provinces)}")
    print(f"First Province: {provinces[0]}")

    # Get statistics
    stats = locations.get_stats()
    print(f"\nStatistics: {stats}")

    # Search example
    results = locations.search('kigali')
    print(f"\nSearch results for 'kigali': {len(results)} results")

File: rwanda-locations.js
/**
 * Rwanda Locations - Easy Integration Library
 *
 * A simple, powerful utility to work with Rwanda's administrative locations.
 * Supports: Provinces, Districts, Sectors, Cells, and Villages
 *
 * For a more feature-rich package version, visit:
 * https://github.com/DevRW/rwanda-location
 */

const fs = require('fs');
const path = require('path');

class RwandaLocations {
  constructor() {
    this.data = null;
    this.cache = {
      provinces: null,
      districts: null,
      sectors: null,
      cells: null,
      villages: null
    };
  }

  /**
   * Load the locations data from JSON file
   */
  load() {
    if (!this.data) {
      const dataPath = path.join(__dirname, 'locations.json');
      this.data = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
    }
    return this;
  }

  /**
   * Get all provinces
   * @returns {Array} List of unique provinces
   */
  getProvinces() {
    if (this.cache.provinces) return this.cache.provinces;

    this.load();
    const provinces = new Map();

    this.data.forEach(location => {
      if (!provinces.has(location.province_code)) {
        provinces.set(location.province_code, {
          code: location.province_code,
          name: location.province_name
        });
      }
    });

    this.cache.provinces = Array.from(provinces.values()).sort((a, b) => a.code - b.code);
    return this.cache.provinces;
  }

  /**
   * Get all districts, optionally filtered by province
   * @param {number|string} provinceCode - Optional province code to filter by
   * @returns {Array} List of districts
   */
  getDistricts(provinceCode = null) {
    this.load();
    const districts = new Map();

    this.data.forEach(location => {
      if (provinceCode && location.province_code != provinceCode) return;

      if (!districts.has(location.district_code)) {
        districts.set(location.district_code, {
          code: location.district_code,
          name: location.district_name,
          province_code: location.province_code,
          province_name: location.province_name
        });
      }
    });

    return Array.from(districts.values()).sort((a, b) => a.code - b.code);
  }

  /**
   * Get all sectors, optionally filtered by district
   * @param {number|string} districtCode - Optional district code to filter by
   * @returns {Array} List of sectors
   */
  getSectors(districtCode = null) {
    this.load();
    const sectors = new Map();

    this.data.forEach(location => {
      if (districtCode && location.district_code != districtCode) return;

      if (!sectors.has(location.sector_code)) {
        sectors.set(location.sector_code, {
          code: location.sector_code,
          name: location.sector_name,
          district_code: location.district_code,
          district_name: location.district_name,
          province_code: location.province_code,
          province_name: location.province_name
        });
      }
    });

    return Array.from(sectors.values()).sort((a, b) => a.code.localeCompare(b.code));
  }

  /**
   * Get all cells, optionally filtered by sector
   * @param {string} sectorCode - Optional sector code to filter by
   * @returns {Array} List of cells
   */
  getCells(sectorCode = null) {
    this.load();
    const cells = new Map();

    this.data.forEach(location => {
      if (sectorCode && location.sector_code != sectorCode) return;

      if (!cells.has(location.cell_code)) {
        cells.set(location.cell_code, {
          code: location.cell_code,
          name: location.cell_name,
          sector_code: location.sector_code,
          sector_name: location.sector_name,
          district_code: location.district_code,
          district_name: location.district_name,
          province_code: location.province_code,
          province_name: location.province_name
        });
      }
    });

    return Array.from(cells.values()).sort((a, b) => a.code - b.code);
  }

  /**
   * Get all villages, optionally filtered by cell
   * @param {number|string} cellCode - Optional cell code to filter by
   * @returns {Array} List of villages
   */
  getVillages(cellCode = null) {
    this.load();
    const villages = new Map();

    this.data.forEach(location => {
      if (cellCode && location.cell_code != cellCode) return;

      if (!villages.has(location.village_code)) {
        villages.set(location.village_code, {
          code: location.village_code,
          name: location.village_name,
          cell_code: location.cell_code,
          cell_name: location.cell_name,
          sector_code: location.sector_code,
          sector_name: location.sector_name,
          district_code: location.district_code,
          district_name: location.district_name,
          province_code: location.province_code,
          province_name: location.province_name
        });
      }
    });

    return Array.from(villages.values()).sort((a, b) => a.code - b.code);
  }

  /**
   * Search locations by name (case-insensitive)
   * @param {string} searchTerm - Search term
   * @param {string} level - Level to search: 'province', 'district', 'sector', 'cell', 'village', or 'all'
   * @returns {Array} Matching locations
   */
  search(searchTerm, level = 'all') {
    this.load();
    const term = searchTerm.toLowerCase();
    const results = [];
    const seen = new Set();

    this.data.forEach(location => {
      const matches =
        (level === 'all' || level === 'province') && location.province_name.toLowerCase().includes(term) ||
        (level === 'all' || level === 'district') && location.district_name.toLowerCase().includes(term) ||
        (level === 'all' || level === 'sector') && location.sector_name.toLowerCase().includes(term) ||
        (level === 'all' || level === 'cell') && location.cell_name.toLowerCase().includes(term) ||
        (level === 'all' || level === 'village') && location.village_name.toLowerCase().includes(term);

      if (matches) {
        const key = `${location.province_code}-${location.district_code}-${location.sector_code}-${location.cell_code}-${location.village_code}`;
        if (!seen.has(key)) {
          seen.add(key);
          results.push({
            province: { code: location.province_code, name: location.province_name },
            district: { code: location.district_code, name: location.district_name },
            sector: { code: location.sector_code, name: location.sector_name },
            cell: { code: location.cell_code, name: location.cell_name },
            village: { code: location.village_code, name: location.village_name }
          });
        }
      }
    });

    return results;
  }

  /**
   * Get location hierarchy for a specific village
   * @param {number|string} villageCode - Village code
   * @returns {Object|null} Complete location hierarchy
   */
  getLocationByVillageCode(villageCode) {
    this.load();
    const location = this.data.find(loc => loc.village_code == villageCode);

    if (!location) return null;

    return {
      country: { code: location.country_code, name: location.country_name },
      province: { code: location.province_code, name: location.province_name },
      district: { code: location.district_code, name: location.district_name },
      sector: { code: location.sector_code, name: location.sector_name },
      cell: { code: location.cell_code, name: location.cell_name },
      village: { code: location.village_code, name: location.village_name }
    };
  }

  /**
   * Get hierarchy tree structure
   * @param {Object} options - Options for filtering (provinceCode, districtCode, etc.)
   * @returns {Array} Hierarchical tree structure
   */
  getHierarchy(options = {}) {
    this.load();
    const { provinceCode, districtCode, sectorCode, cellCode } = options;

    const tree = {};

    this.data.forEach(location => {
      if (provinceCode && location.province_code != provinceCode) return;
      if (districtCode && location.district_code != districtCode) return;
      if (sectorCode && location.sector_code != sectorCode) return;
      if (cellCode && location.cell_code != cellCode) return;

      const pKey = location.province_code;
      const dKey = location.district_code;
      const sKey = location.sector_code;
      const cKey = location.cell_code;

      if (!tree[pKey]) {
        tree[pKey] = {
          code: location.province_code,
          name: location.province_name,
          districts: {}
        };
      }

      if (!tree[pKey].districts[dKey]) {
        tree[pKey].districts[dKey] = {
          code: location.district_code,
          name: location.district_name,
          sectors: {}
        };
      }

      if (!tree[pKey].districts[dKey].sectors[sKey]) {
        tree[pKey].districts[dKey].sectors[sKey] = {
          code: location.sector_code,
          name: location.sector_name,
          cells: {}
        };
      }

      if (!tree[pKey].districts[dKey].sectors[sKey].cells[cKey]) {
        tree[pKey].districts[dKey].sectors[sKey].cells[cKey] = {
          code: location.cell_code,
          name: location.cell_name,
          villages: []
        };
      }

      const villages = tree[pKey].districts[dKey].sectors[sKey].cells[cKey].villages;
      if (!villages.find(v => v.code === location.village_code)) {
        villages.push({
          code: location.village_code,
          name: location.village_name
        });
      }
    });

    // Convert nested objects to arrays
    return Object.values(tree).map(province => ({
      ...province,
      districts: Object.values(province.districts).map(district => ({
        ...district,
        sectors: Object.values(district.sectors).map(sector => ({
          ...sector,
          cells: Object.values(sector.cells)
        }))
      }))
    }));
  }

  /**
   * Get statistics about the data
   * @returns {Object} Statistics
   */
  getStats() {
    this.load();

    return {
      totalLocations: this.data.length,
      provinces: this.getProvinces().length,
      districts: this.getDistricts().length,
      sectors: this.getSectors().length,
      cells: this.getCells().length,
      villages: this.getVillages().length
    };
  }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = RwandaLocations;
}

// Export for browser (if needed)
if (typeof window !== 'undefined') {
  window.RwandaLocations = RwandaLocations;
}
